### PR TITLE
[Exam] Test Run follow-up

### PR DESCRIPTION
**Pull request by anditurdiu**
# Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
As a follow-up to #2096 this PR addresses:
- REST API: Change names of rest call for delete & create test run 
- Currently, the test run exercises are ordered by the `id` of the exercise groups. They should follow the order of the exercises groups, as configured for the exam.
- Empty exercise groups do not allow the creation of test runs. They should be ignored when creating a test run. 
- Reduce Code duplication
- Improve the performance, complexity and readability of `studentExamService.deleteTestRun()`

**Model Changes:**
- Add `started_date` to `student_exam`. This should be set only when the student first starts the exam conduction.
- Change `last_modified_date`, `created_date` and `submission_date` datatype of `student_exam` to `TIMESTAMP`
- Change `last_modified_date` `created_date` datatype of `exam_session` to `TIMESTAMP`

### Description

- Change Routes to:
> DELETE courses/{courseId}/exams/{examId}/test-run/{testRunId}
> POST courses/{courseId}/exams/{examId}/test-run/
- Lock the create test run button if no exercises are found
- Ignore exercise groups with no exercises in the test run modal
- Keep the order of the exercise groups for the test runs
- Refactor the deleteTestRun() to reduce complexity and improve performance
> Fetch only relevant exercise participations (no need to filter after)
- Improve the client test and server test coverage

### Steps for Testing
1. Log in to Artemis
2. Create an exam with 2 exercise groups and no exercises
3. Navigate to the test runs, you should see the locked button and the tooltip for `Create Test Run` 
4. Add an exercise to one of the exercise groups
5. Navigate to the test runs, create a test run with one exercise.
6. Start the test run and finish it
7. Add an exercise to the other exercise group
8. Change the order of the exercise groups
9. Create a new test run. 
10. Check the order of the exercises, it should follow the order set for the exercise groups

### Test Coverage
**Client:**

- create-test-run-modal.component.ts 64% -> 70.59%
- test-run-management.component.ts 77% -> 79.03%

**Server:**

- StudentExamService: 75% -> 82%

### Screenshots
Added Tooltips when disabling buttons: 
image
image

### CODE REVIEW

**Review by stefanwaldhauser — APPROVED** [review]
Works as expected! Nide OWrk

File: src/main/webapp/app/exam/manage/test-runs/test-run-management.component.ts [hunk 1]
```diff
@@ -128,6 +128,19 @@ export class TestRunManagementComponent implements OnInit {
         return false;
     }
 
+    get examContainsExercises(): boolean {
```
Comment (TobiasPr) [inline src/main/webapp/app/exam/manage/test-runs/test-run-management.component.ts:131]: Could be shortened: `this.exam.exerciseGroups.some(exerciseGroup => exerciseGroup.exercises && exerciseGroup.exercises.length >  0)`

**Review by TobiasPr — CHANGES_REQUESTED** [review]
Please fix the mention of github bot in src/test/javascript/spec/component/exam/test-run/create-test-run-modal.component.spec.ts (unused import)
Consider shortening examContainsExercises in src/main/webapp/app/exam/manage/test-runs/test-run-management.component.ts for better readability

File: src/main/webapp/app/exam/manage/test-runs/test-run-management.component.ts (hunk 1 above)
Comment (anditurdiu) [reply src/main/webapp/app/exam/manage/test-runs/test-run-management.component.ts:134]: Yeah, nice suggestion. I implemented it accordingly

**Review by stefanwaldhauser — APPROVED** [review]
reapprove

**Review by JonasPetry — APPROVED** [review]
Did not check the code, but functionality works as you describe. Nice work 🚀

**Review by kloessst — CHANGES_REQUESTED** [review]
I get some console errors when I try to follow the testing steps. I had one exerciseGroup with an exercise and one without one.

consoleErrors

**Review by fde312 — CHANGES_REQUESTED** [review]
When no exerciseGroups are defined (see step 3 testing steps)
image

After adding exercise group and exercise still:
image

I get also an error when trying to create a test run and the test run is not created:
image

**Comment by anditurdiu** [comment]
> When no exerciseGroups are defined (see step 3 testing steps)
> image
> 
> After adding exercise group and exercise still:
> image
> 
> I get also an error when trying to create a test run and the test run is not created:
> image
> 
> I was able to create a test run after trying for the second time but I still get the console errors:
> image
> 
> Has it something todo that we deployed on ts3? For ts1 it seemed to work (see Jonas approve)

@kloessst  @fde312 I just pushed a fix, it was all due to a missing `?`, see 98d1cd33762dfd71afff7aa6e690b8e7f385a54b
No console errors should be thrown anymore

**Review by TobiasPr — APPROVED** [review]
Works for me, good job :)

File: src/main/resources/config/liquibase/changelog/20200925174332_changelog.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.9.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.9.xsd">
+<changeSet author="turdiu" id="20200925174332">
```
Comment (derLalla) [inline src/main/resources/config/liquibase/changelog/20200925174332_changelog.xml:3]: is it necessary to upload the changelog?

File: src/main/webapp/i18n/de/exam.json [hunk 3]
```diff
@@ -204,10 +204,12 @@
...
+                "setupHint": "Wähle für jede Aufgabengruppe eine Aufgabe aus, indem du auf die Aufgabe klickst.\n<strong>Hinweis:</strong> Wenn du bereits einen Testlauf mit einer der Aufgaben erstellt hast, dann wird diese bereits eine Abgabe und Ergebnis enthalten.",
                 "assess": "Deine Testläufe Bewerten",
                 "manage": "Arbeitszeit verwalten",
                 "text": "Hier kannst du einen Testlauf für die Klausur erstellen. Wähle aus, welche Aufgaben die Klausur enthalten soll und führe sie aus. Anschliessend, kannst du die Abgabe als Tutor bewerten. Das Ergebnis, kann du als Student einsehen und dich Beschweren. Diese Beschwerden kannst du dann als Tutor bewerten.\n<strong>Hinweis:</strong> Abgaben innerhalb von testläufen werden nicht für die Prüfungsergebnisse gezählt.",
+                "createTestRunDisabled": "Keine Aufgaben gefunden",
+                "assessTestRunDisabled": "Keins deiner Testläufe ist beended",
```
Comment (derLalla) [inline src/main/webapp/i18n/de/exam.json:212]: typo: Keiner deiner Testläufe ist beendet

Comment (anditurdiu) [reply src/main/webapp/i18n/de/exam.json]: Thanks, its fixed here: 47726c02c0873a7fc17eb3b0deebbf7359982629

**Review by stefanwaldhauser — APPROVED** [review]
reapprove

**Comment by anditurdiu** [comment]
> still one typo at the end: "beende**t**"

Fix here: d1090042bc4a3686b2842ced0aeddc0dc7bcbf1c
When you are reviewing the code especially for Typos it is easier if you directly insert the suggestion like so: 
image
This is very helpful especially for us non-mothertounge speakers, instead of going back and forth with commits :D

**Review by TobiasPr — DISMISSED** [review]
Reapprove -> fixed typo

File: src/main/webapp/i18n/de/exam.json [hunk 4]
```diff
@@ -204,10 +204,12 @@
...
                 "assess": "Deine Testläufe Bewerten",
                 "manage": "Arbeitszeit verwalten",
-                "text": "Hier kannst du einen Testlauf für die Klausur erstellen. Wähle aus, welche Aufgaben die Klausur enthalten soll und führe sie aus. Anschliessend, kannst du die Abgabe als Tutor bewerten. Das Ergebnis, kann du als Student einsehen und dich Beschweren. Diese Beschwerden kannst du dann als Tutor bewerten.\n<strong>Hinweis:</strong> Abgaben innerhalb von testläufen werden nicht für die Prüfungsergebnisse gezählt.",
+                "text": "Hier kannst du einen Testlauf für die Klausur erstellen. Wähle aus, welche Aufgaben die Klausur enthalten soll und führe sie aus. Anschliessend, kannst du die Abgabe als Tutor bewerten. Das Ergebnis, kann du als Student einsehen und dich Beschweren. Diese Beschwerden kannst du dann als Tutor bewerten.\n<strong>Hinweis:</strong> Abgaben innerhalb von Testläufen werden nicht bei den Klausurergebnissen gezählt.",
```
Comment (derLalla) [inline src/main/webapp/i18n/de/exam.json:210]:
```suggestion
                "text": "Hier kannst du einen Testlauf für die Klausur erstellen. Wähle aus, welche Aufgaben die Klausur enthalten soll und führe sie aus. Anschließend, kannst du die Abgabe als Tutor bewerten. Das Ergebnis, kannst du als Student einsehen und dich beschweren. Diese Beschwerden kannst du dann als Tutor bewerten.\n<strong>Hinweis:</strong> Abgaben innerhalb von Testläufen werden nicht bei den Klausurergebnissen gezählt.",
```

**Review by derLalla — DISMISSED** [review]

**Comment by derLalla** [comment]
> This is very helpful especially for us non-native speakers, instead of going back and forth with commits :D
Yes you`re right! Thanks for mentioning :)

File: src/main/webapp/i18n/de/exam.json (hunk 4 above)
Comment (anditurdiu) [reply src/main/webapp/i18n/de/exam.json]: Done a42fae5162bf909dfdaa3e7857aaf612ac9984e4

**Review by stefanwaldhauser — APPROVED** [review]
Reapprove

**Review by fde312 — APPROVED** [review]
Nullpointer is solved, and now it works fine for me. Code looks good too! Great job! 🚀

**Review by filip-gregurevic — APPROVED** [review]

**Review by kloessst — APPROVED** [review]
Works fine now. Code also looks good!

**Review by jpbernius — APPROVED** [review]

**Review by krusche — APPROVED** [review]

**Pull request merged**